xep-0388.xml

```diff
@@ -325,11 +325,11 @@
   </section2>
   
   <section2 topic="SASL mechanism upgrade tasks" anchor="upgrade">
-    <p>In order to upgrade the Server's database to include the needed data for new SASL mechanisms, the client indicated the desired SASL mechanism to upgrade in its &lt;authenticate/> element using the "upgrade" attribute as specified in <link url="#initiation">Initiation</link>. The server provides a task to perform this upgrade. This task MUST be the only task provided in this &lt;continue/> element and, for security reasons, the upgrade task MUST be the last task the server requires.
-    This task has a prefix of "UPGR-" followed by the SASL mechanism name to upgrade to.</p>
-    <p>The client then initiates this upgrade task by providing a corresponding &lt;next/> element providing the task name.
-    Upon receiving the &lt;next/> element for the upgrade, the server provides the elements and data needed for the client to calculate the requested data. The concrete elements and exchanges needed for the upgrade task have to be defined in a future protocol.</p>
-    <p>If the SASL mechanism to upgrade to supports it, no password SHOULD be send in clear.</p>
+    <p>Modern XMPP server deployments typically store only the hash of a user's password, to improve account security. At times, it may be desirable for servers to upgrade to newer or different hash algorithms, e.g. so they can offer different authentication mechanisms for improved security or interoperability. Due to the security properties of one-way cryptographic hash algorithms, it is not possible for the server to extract the original data and simply hash it in a new format. To perform such upgrades, the cooperation of the client is necessary - because it has, or can obtain from the user, the original password to derive a hash from.</p>
+    <p>As described in the <link url="#feature">SASL mechanism negotiation</link> section, the server can inform the client of one or more upgrades that need to be performed for the current account. The client SHOULD always execute any upgrade task it recognises. To do this, it MUST include the 'upgrade' attribute of its &lt;authenticate/> element, as specified in <link url="#initiation">Initiation</link>.</p>
+    <p>Upon successfully authenticating the client (including any secondary authentication steps required for the account), but before the final &lt;success/> would be sent, the server sends a &lt;continue/> element, which MUST contain a single task, matching whatever was selected by the client. The client then initiates this upgrade task by providing a corresponding &lt;next/> element providing the task name.</p>
```

Comment:
```suggestion
    <p>Upon successfully authenticating the client (including any secondary authentication steps required for the account), but before the final &lt;success/> would be sent, the server sends a &lt;continue/> element, which MUST contain a single upgrade task, matching whatever was selected by the client. The client then initiates this upgrade task by providing a corresponding &lt;next/> element providing this task name.</p>
```

Reply:
What is the intent of this suggested change? The text was aiming to preserve the requirements of this replaced text:

> This task MUST be the only task provided in this &lt;continue/> element 

I think the suggested change doesn't preserve this, as it can easily be read as "a single upgrade task" but any number of other tasks as well.

Reply:
I did not want to change any meaning, but for me, my suggested changes make it even more clear...but if you think that my changes rather do the opposite, just leave them out :)